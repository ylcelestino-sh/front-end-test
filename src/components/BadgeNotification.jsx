import React from 'react'
import PropTypes from "prop-types";
import { FiShoppingCart } from 'react-icons/fi';

import { Box, Button } from '@chakra-ui/react';

const BadgeNotification = ({itemsNumbers = 0}) => {
  return (
    <Button py={"2"} aria-label={"Notifications"} variant="ghost">
    <>
      <FiShoppingCart />
      <Box
        as={"span"}
        color={"white"}
        position={"absolute"}
        top={"6px"}
        right={"4px"}
        w="30%"
        fontSize={"0.8rem"}
        bgColor={"red"}
        borderRadius="3xl"
        zIndex={9999}
        p={"1px"}
        variant="ghost"
      >
        {itemsNumbers}
      </Box>
    </>
  </Button>
  )
}

BadgeNotification.propTypes = {
    itemsNumbers: PropTypes.number
}
export default BadgeNotification;
