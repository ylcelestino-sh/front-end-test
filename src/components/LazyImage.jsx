import React from "react";
import PropTypes from "prop-types";
import { Image, Spinner } from "@chakra-ui/react";
import useLazyLoadImage from "../hooks/useLazyLoadImage";
import { PLACEHOLDER } from "../constant/constant";

const LazyImage = ({ alt, src, rounded='lg', height=150, width=282, objectFit='contain' }) => {
  const { imageSrc, setImageRef, isLoading } = useLazyLoadImage(src);
  return (
    <>
        <Image
          data-testid="lazyImage"
          ref={setImageRef}
          src={imageSrc}
          alt={alt}
          rounded={rounded}
          height={height}
          width={width}
          objectFit={objectFit}
          fallbackSrc={PLACEHOLDER}
        />
      {
        isLoading ? <Spinner /> : null
      }
    </>
  );
};

LazyImage.propTypes = {
  src: PropTypes.string,
  alt: PropTypes.string,
  rounded: PropTypes.string,
  height: PropTypes.number,
  width: PropTypes.number,
  objectFit: PropTypes.string,
};

export default LazyImage;
