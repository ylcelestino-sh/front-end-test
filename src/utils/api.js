import axios from 'axios';

class ApiService {

  constructor() {
    this.baseUrl = process.env.REACT_APP_API_URL || 'https://front-test-api.herokuapp.com/';
    this.axiosInstance = axios.create({
      baseURL: this.baseUrl,
      headers: this.getHeaders(),
    });
  }
  // get Headers for create axios instance
  getHeaders() {
    const headers = {
      'content-type': 'application/json',
    };
    return headers;
  }
}
export default ApiService;