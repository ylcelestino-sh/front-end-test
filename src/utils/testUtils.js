import React from "react";
import PropTypes from "prop-types";
import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { setupStore } from "../app/store";
import { ChakraProvider } from "@chakra-ui/react";
import { BrowserRouter as Router } from "react-router-dom";

/*
 * Modified version of RTK Query's helper function:
 * https://github.com/reduxjs/redux-toolkit/blob/master/packages/toolkit/src/query/tests/helpers.tsx
 */

const AllProviders = ({ children }) => {
  AllProviders.propTypes = {
    children: PropTypes.node,
  };
  return (
    <Provider store={setupStore()}>
      <ChakraProvider>
        <Router>{children}</Router>
      </ChakraProvider>
    </Provider>
  );
};

export function withProvider(store) {
  return function Wrapper({ children }) {
    Wrapper.propTypes = {
      children: PropTypes.node,
    };
    return <Provider store={store}>{children}</Provider>
  }
}

const renderWithProviders = (
  ui,
  { route = "/products" } = {},
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) => {
  renderWithProviders.propTypes = {
    ui: PropTypes.node,
  };
  window.history.pushState({}, "Test page", route);

  return {
    store,
    AllProviders,
    ...render(ui, { wrapper: AllProviders, ...renderOptions }),
  };
};

const getStore = ({api, extraReducers}) => {
  return configureStore({
    reducer: combineReducers({
      [api.reducerPath]: api.reducer,
      ...extraReducers,
    }),
    middleware: (gdm) =>
      gdm({ serializableCheck: false, immutableCheck: false }).concat(
        api.middleware
      ),
  });
};

const setupApiStore = (api, extraReducers = {}) => {
  const initialStore = getStore({ api, extraReducers });
  const refObj = {
    api,
    store: initialStore,
    wrapper: withProvider({initialStore}),
  };

  const store = getStore({ api, extraReducers });
  refObj.store = store;

  return refObj;
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { renderWithProviders as render, setupApiStore };
