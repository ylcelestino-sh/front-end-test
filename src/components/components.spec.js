import React from "react";
import { PLACEHOLDER } from "../constant/constant";
import { render, screen } from "../utils/testUtils";
import BadgeNotification from "./BadgeNotification";
import LazyImage from "./LazyImage";
describe("custom components test", () => {
  // BadgeNotification
  describe("BadgeNotification test", () => {
    test("BadgeNotification should show correctly with props", () => {
      render(<BadgeNotification itemsNumbers={1} />);
      const badgeNotificationItem = screen.getByTestId("itemNumber");

      expect(badgeNotificationItem).toMatchSnapshot();
      expect(badgeNotificationItem.textContent).toBe("1");
    });
    test("BadgeNotification should show correctly without props", () => {
      render(<BadgeNotification />);
      const badgeNotificationItem = screen.getByTestId("itemNumber");

      expect(badgeNotificationItem).toMatchSnapshot();
      expect(badgeNotificationItem.textContent).toBe("0");
    });
  });

  // LazyImage
  describe("LazyImage test", () => {
    test("LazyImage should show correctly", () => {
      render(<LazyImage alt={"fake alt"} src={PLACEHOLDER} />);
      expect(screen.getByTestId("lazyImage")).toMatchSnapshot();
    });

    test.each(["", null, undefined])(
      "LazyImage should show default placeholder when src it's empty, null or undefined",
      (src) => {
        render(<LazyImage alt={"fake alt"} src={src} />);
        expect(
          screen.getByTestId("lazyImage").attributes.getNamedItem("src").value
        ).toBe(PLACEHOLDER);
      }
    );
  });
});
