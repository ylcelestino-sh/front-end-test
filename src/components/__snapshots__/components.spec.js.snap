// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom components test BadgeNotification test BadgeNotification should show correctly with props 1`] = `
<span
  class="css-1szyxpv"
  data-testid="itemNumber"
  variant="ghost"
>
  1
</span>
`;

exports[`custom components test BadgeNotification test BadgeNotification should show correctly without props 1`] = `
<span
  class="css-1szyxpv"
  data-testid="itemNumber"
  variant="ghost"
>
  0
</span>
`;

exports[`custom components test LazyImage test LazyImage should show correctly 1`] = `
<img
  alt="fake alt"
  class="chakra-image__placeholder css-zv8j3m"
  data-testid="lazyImage"
  src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII="
/>
`;
